pennylane/devices/execution_config.py

```diff
@@ -76,12 +88,49 @@ def __post_init__(self):
             )
 
         if self.device_options is None:
-            self.device_options = {}
+            object.__setattr__(self, "device_options", frozendict({}))
 
         if self.gradient_keyword_arguments is None:
-            self.gradient_keyword_arguments = {}
+            object.__setattr__(self, "gradient_keyword_arguments", frozendict({}))
 
         if any(arg not in SUPPORTED_GRADIENT_KWARGS for arg in self.gradient_keyword_arguments):
             raise ValueError(
                 f"All gradient_keyword_arguments keys must be in {SUPPORTED_GRADIENT_KWARGS}, got unexpected values: {set(self.gradient_keyword_arguments) - set(SUPPORTED_GRADIENT_KWARGS)}"
             )
+
+        # Initialize private shot attributes
+        shots, shot_vector, raw_shot_sequence = None, None, None
+
+        if self.shots is None:
+            # analytic mode
+            shots = self.shots
+            shot_vector = None
+            raw_shot_sequence = None
+
+        elif isinstance(self.shots, int):
+            # sampling mode (unbatched)
+            if self.shots < 1:
+                raise DeviceError(
+                    f"The specified number of shots needs to be at least 1. Got {self.shots}."
+                )
+
+            shots = self.shots
+            shot_vector = None
+
+        elif isinstance(self.shots, Sequence) and not isinstance(self.shots, str):
+            # batched sampling mode
+            shots, shot_vector = _process_shot_sequence(self.shots)
+            raw_shot_sequence = self.shots
+
+        else:
+            raise DeviceError(
+                "Shots must be a single non-negative integer or a sequence of non-negative integers."
+            )
+
+        object.__setattr__(self, "_shots", shots)
+        object.__setattr__(self, "shots", shots)
```

Comment:
Why not `self._shots = shots`? And why do we want both public and private attributes?